### PR TITLE
Reseed all Generators in Dataloader's _worker_loop()

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR addresses https://fb.workplace.com/groups/pytorch.oss.dev/posts/1699944830430051 and does a bunch of stacked changes:

1. Make `Generator` weakref-able (C++ part)
2. Create a registry of manually-created `Generator` objects (via weakrefs)
3. Use this registry in the `Dataloader`'s `_worker_loop` to re-seed all existing `Generator` instances: this extends what is already applied to the global `Generator`, which is already re-seeded.

CC @albanD @ezyang as previously discussed.

TODO: a bit of docs and justification, which I'll do if this PR is mergeable.

cc @SsnL @VitalyFedyunin @ejguan @dzhulgakov @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @pbelevich

### CODE REVIEW

**Comment by NicolasHug** [comment]
Closing for https://github.com/pytorch/pytorch/pull/107034, sorry for the noise
